### ISSUE TITLE
SELECT unicode string not work

### ISSUE THREAD

**Issue by kennic** [ccgus/fmdb#391]
I can't use SELECT with unicode string.
For example:
I have string "tést" in table.
SELECT string FROM table WHERE string="test"
-> it returned empty.
How can I still get "tést" when I query with "test" ?

**Comment by robertmryan** [comment]
I don't know if you're using Objective-C or Swift, but here's a function you can use to remove diacritics from strings: http://stackoverflow.com/a/27576609/1271826

**Comment by kennic** [comment]
Here's my objective-c version:

```objective-c
[db makeFunctionNamed:@"unaccented" maximumArguments:1 withBlock:^(sqlite3_context *context, int argc, sqlite3_value **argv)
{
    const unsigned char *utf8String = sqlite3_value_text(argv[0]);
    NSMutableString *string = [NSMutableString stringWithUTF8String:(const char*)utf8String];
    CFStringTransform((__bridge CFMutableStringRef)string , nil, kCFStringTransformStripCombiningMarks, 0);
    sqlite3_result_text(context, string.UTF8String, -1, SQLITE_STATIC);
}];
```
usage:
SELECT * FROM strings WHERE unaccented(string)='whatever'
it works, but the performance is slow.

**Comment by billymeltdown** [comment]
> On Jul 22, 2015, at 10:06 AM, Nam Kennic <notifications@github.com> wrote:
> 
> Here's my objective-c version:
> 
> [db makeFunctionNamed:@"unaccented" maximumArguments:1 withBlock:^(sqlite3_context *context, int argc, sqlite3_value **argv)
> {
>     const unsigned char *utf8String = sqlite3_value_text(argv[0]);
>     NSMutableString *string = [NSMutableString stringWithUTF8String:(const char*)utf8String];
>     CFStringTransform((__bridge CFMutableStringRef)string , nil, kCFStringTransformStripCombiningMarks, 0);
>     sqlite3_result_text(context, string.UTF8String, -1, SQLITE_STATIC);
> }];
> usage:
> SELECT * FROM strings WHERE unaccented(string)='whatever'
> it works, but the performance is slow.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/ccgus/fmdb/issues/391#issuecomment-123734063>.
> 

That’s a clever solution, I like it. Are you certain that your function is the performance issue? A full table scan is always expensive. I’d be curious to see what the EXPLAIN plan looks like with such a function and query on your table. For instance:

EXPLAIN SELECT * FROM strings WHERE unaccented(string)=‘whatever’;

Cheers,
Billy

**Comment by robertmryan** [comment]
Absolutely. In my big tables, I create a new "search" column and do the unaccent process on updates and inserts, then I can do quick searches without re-running the unaccent function on every row repeatedly. It's kludgy (much like many denormalization processes), but gives you good performance.

**Comment by robertmryan** [comment]
By the way, you probably want to use an autoreleasepool inside this unaccent function, or else you'll have memory problems in large queries.

**Comment by robertmryan** [comment]
By the way, you probably want to use an `autoreleasepool` inside this `unaccented` function, or else you'll have memory problems in large queries. Also, I think you want `SQLITE_TRANSIENT`, not `SQLITE_STATIC`. So, perhaps something like:

    [db makeFunctionNamed:@"unaccented" maximumArguments:1 withBlock:^(sqlite3_context *context, int argc, sqlite3_value **argv) {
        const unsigned char *utf8String = sqlite3_value_text(argv[0]);
        if (!utf8String) {
            sqlite3_result_null(context);
            return;
        }
        @autoreleasepool {
            NSMutableString *string = [NSMutableString stringWithUTF8String:(const char*)utf8String];
            CFStringTransform((__bridge CFMutableStringRef)string , nil, kCFStringTransformStripCombiningMarks, 0);
            sqlite3_result_text(context, string.UTF8String, -1, SQLITE_TRANSIENT);
        }
    }];

**Comment by billymeltdown** [comment]
I was going to propose that one could create a materialized view of the data you want to search like this, with the string column already `unaccented()` (and perhaps `LOWER()’d` as well). And then I remembered SQLite doesn’t have materialized views. **Sad trombone**. Anyway, you could create a table that caches the data you want to search to avoid the penalty on your search, prepared however you like, and maintained with triggers from the original table on INSERT, UPDATE, and DELETE. Assuming space is no issue!

**Comment by kennic** [comment]
summary for this issue:

```objective-c
[db makeFunctionNamed:@"unaccented" maximumArguments:1 withBlock:^(sqlite3_context *context, int argc, sqlite3_value **argv)
{
    const unsigned char *utf8String = sqlite3_value_text(argv[0]);
    if (!utf8String) {
        sqlite3_result_null(context);
        return;
    }
    @autoreleasepool {
        NSMutableString *string = [NSMutableString stringWithUTF8String:(const char*)utf8String];
        CFStringTransform((__bridge CFMutableStringRef)string , nil, kCFStringTransformStripCombiningMarks, 0);
        sqlite3_result_text(context, string.UTF8String, -1, SQLITE_TRANSIENT);
    }
}];
```
usage:
```sql
SELECT * FROM strings WHERE unaccented(string)=‘whatever’
```
For better performance, create a pre-unaccented string column on INSERT, UPDATE
Thanks for all of your feedback @robertmryan @billymeltdown
